rules-reviewed/camel3/camel2/xml-moved-components.windup.groovy

```diff
@@ -26,13 +27,66 @@ import java.util.function.Function
+    return "//*/c:route/*[starts-with(@uri, '$component:')]"
+}
+
+final Map<String, String[]> movedComponents = [
+        // component name: [xml xpath, java matches expression ]
+        "language"     : [componentNameXpath("language"), componentNameExpression("language")],
+        "bean"         : [componentNameXpath("bean"), componentNameExpression("bean")],
+        "direct"       : [componentNameXpath("direct"), componentNameExpression("direct")],
+        "xslt"         : [componentNameXpath("xslt"), componentNameExpression("xslt")],
+        "browse"       : [componentNameXpath("browse"), componentNameExpression("browse")],
+        "dataset"      : [componentNameXpath("dataset"), componentNameExpression("dataset")],
+        "direct-vm"    : [componentNameXpath("direct-vm"), componentNameExpression("direct-vm")],
+        "file"         : [componentNameXpath("file"), componentNameExpression("file")],
+        "log"          : [componentNameXpath("log"), componentNameExpression("log")],
+        "mock"         : [componentNameXpath("mock"), componentNameExpression("mock")],
+        "ref"          : [componentNameXpath("ref"), componentNameExpression("ref")],
+        "saga"         : [componentNameXpath("saga"), componentNameExpression("saga")],
+        "scheduler"    : [componentNameXpath("scheduler"), componentNameExpression("scheduler")],
+        "seda"         : [componentNameXpath("seda"), componentNameExpression("seda")],
+        "stub"         : [componentNameXpath("stub"), componentNameExpression("stub")],
+        "timer"        : [componentNameXpath("timer"), componentNameExpression("timer")],
+        "validator"    : [componentNameXpath("validator"), componentNameExpression("validator")],
+        "vm"           : [componentNameXpath("vm"), componentNameExpression("vm")],
+        "rest"         : [countXpath("rest", "get", "post", "put", "delete"), "rest({*})"],
+        "xslt "         : [countXpath("xpath"), ".xpath({*})"],
+        "zip-deflater" : [countXpath("zip"), ".zip({*})"],
+        "zip-deflater ": [countXpath("gzip"), ".gzip({*})"],
+        "dataformat "  : [countXpath("marshal,unmarshal,dataformats"), ".marshal({*})"],
+        "dataformat"   : ["", ".unmarshal({*})"],
+        "attachements ": ["", ".addAttachement({*})"]
+]
 
 final IssueCategory optionalIssueCategory = new IssueCategoryRegistry().getByID(IssueCategoryRegistry.OPTIONAL)
 final Link modularizationLink = Link.to("Camel 3 - Migration Guide: Modularization of camel-core", "https://camel.apache.org/manual/latest/camel-3-migration-guide.html#_modularization_of_camel_core")
 
-final BiFunction<String, String, Boolean> isThereArtifactId = { String xmlDependenciesBlock, String component -> 
-    xmlDependenciesBlock.contains("<artifactId>camel-$component</artifactId>" as CharSequence) }
+
+final BiFunction<String, String, Boolean> isThereArtifactId = { String xmlDependenciesBlock, String component ->
```

Comment:
Thank you :) You gave me opportunity to refresh what's a BiFunction and a Function..... it's been a while without seeing it the code :+1: